### ISSUE TITLE
unicode problems in restarting mpi runs

### ISSUE THREAD

**Issue by mgiulini** [haddocking/haddock3#558]
currently the restart option does not allow to restart a haddock3 MPI run, as the following error occurs:
```
  ...
  File "/trinity/login/mgiulini/haddock3/src/haddock/gear/prepare_run.py", line 273, in setup_run
    update_step_contents_to_step_names(
  File "/trinity/login/mgiulini/haddock3/src/haddock/gear/prepare_run.py", line 905, in update_step_contents_to_step_names
    update_step_names_in_file(file_, prev_names, new_names)
  File "/trinity/login/mgiulini/haddock3/src/haddock/gear/prepare_run.py", line 925, in update_step_names_in_file
    text = file_.read_text()
  File "/trinity/login/mgiulini/anaconda3/envs/haddock3/lib/python3.9/pathlib.py", line 1256, in read_text
    return f.read()
  File "/trinity/login/mgiulini/anaconda3/envs/haddock3/lib/python3.9/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte
[2022-09-26 10:03:43,829 libutil ERROR] 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte
```
this occurs because we are trying to read the `mpi.pkl` file, which contains non-ASCII characters. This can probably be fixed by correcting the `encoding` of `file_.read_text()`.
I have another doubt: the goal of this part of the code is, as far as I understand, to replace the old module names with the new ones. But apparently this is done even if the names are equal. Should we correct this?
@joaomcteixeira maybe we can use your help here :)

**Comment by joaomcteixeira** [comment]
Hi,

> Should we correct this?

No. It is done that way on purpose to update the integer prefixes of the step folders. For example, if the original run had 5 steps, the folders are `0_`, `1_`, etc. BUT if the new run with `--restart` has 12 steps, the prefix must be updated to `00_`, `01_`, etc.

**IMPORTANT** these names must also be updated inside the files; that is, the file contents. There's where the error reported here is happening.

**Easy fix:** Does the `mpi.pkl` contain any crucial information on the name of the original step folders? Or is it a compiled file? If the answer is no, you can make a `try:except` block that, in case of an error reading the file, copy the file as is (from within the `except` part.

**Less easy fix:** as you said. Try reading the non-ASCII files with another type of encoding and replacing the strings. OR read them in bytes ([see here](https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_bytes)) and replace the bytes directly instead of working with strings. 

Cheers

**Comment by joaomcteixeira** [comment]
> here we are still calling the replace function even if the two names (s1 and s2) coincide: I would correct this by introducing an if clause, do you agree?

It doesn't matter. Sometimes I think like the following: "Is the process time-consuming or creating any side effects? If the answer is "no", why make an `if`?" In this particular case, I directly use the `replace`, if there's nothing to do, nothing will be done. The computer cost for this particular operation is even faster than the `if`, I think. Because with the `if`, you have to assert `if substring in the string` and then if it exists, you need to replace; which is another find event. With the `replace`, you search the string only once and the `if` happens internally.

In conclusion, there's nothing wrong with placing an `if` there. But in this case, it will only make the code more complicated without saving any time; it will likely cost more time, actually.

Cheers :wink:

**Comment by mgiulini** [comment]
I think I do not agree with this..this renaming process can be quite time-consuming when you have large runs (imagine restarting from step 8 of a run with 10k rigidbody sampling and 5k flexref structures). And it has the other disadvantage of opening and writing to disk a huge amount of files. And I guess that in the vast majority of the cases the `prev_names` and `new_names` lists will be exactly equal making this process is not necessary.

I understand your point of view on the simplicity, but I would say that the computational time that can be saved in this context is worth the a new if clause here 
https://github.com/haddocking/haddock3/blob/b69f6e3dcea9b10ec1a7add1d5b7f9dab72e7831/src/haddock/gear/prepare_run.py#L896-L897

do you agree?

**Comment by joaomcteixeira** [comment]
I see what you mean; you are talking on the more general view instead of the "per-file" operation. This operation will indeed be performed on every file and in any situation where the `--restart` is used. So, yes, you can make a check to assert if there is a need to change the files or not before entering the loop. 

In that case, I would rather add a function here:

https://github.com/haddocking/haddock3/blob/b69f6e3dcea9b10ec1a7add1d5b7f9dab72e7831/src/haddock/gear/prepare_run.py#L273

I think this can be done if `prev_` and `new_` are different. For example:

```python
_prev, _new = renum_step_folders(general_params[RUNDIR])
if _prev != _new:
    update_step_contents_to_step_names(...)
```

If `_prev` and `_new` are equal I think there's no need to change the files either. I hope I am not missing any detail.

Good point!

**Comment by mgiulini** [comment]
Ok, thanks for the clarification! so as far as the restart option is concerned, the module-specific `mpi.pkl` can be safely ignored I guess (these unpacked modules will never need to be executed again)

**Comment by joaomcteixeira** [comment]
> This is very unnecessarily complex, just set the initial number to a high prefix: 00001_ etc, should save some future troubles

FYI: In the beginning, it was like that. But, with the work and feedback from @douweschulte we made the numbers vary to enhance the user experience. See https://github.com/haddocking/haddock3/pull/378 and associated PRs/issues for more info.

**Comment by amjjbonvin** [comment]
I would think that a 00X_, 0XX_ notation is rather simple and could simplify the machinery (if needed).

As long as the directories are sorted logically when making a listing the user won’t bother I would think.

I don’t think that there will be workflows that exceed a few tens of steps. With three digits we can go to 999 steps.


> FYI: In the beginning, it was like that. But, with the work and feedback from @douweschulte <https://github.com/douweschulte> we made the numbers vary to enhance the user experience. See #378 <https://github.com/haddocking/haddock3/pull/378> and associated PRs/issues for more info.

**Comment by mgiulini** [comment]
same issue occurs with the `haddock-copy` cli
